core/src/main/java/org/elasticsearch/common/xcontent/ObjectSpooler.java

```diff
@@ -0,0 +1,301 @@
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import static java.lang.Math.min;
+
+/**
+ * Like {@link ObjectParser} but works with objects that have constructors who's arguments are mixed in with its other settings. Queries are
+ * like this, for example <code>ids</code> requires <code>types</code> but always parses the <code>values</code> field on the same level. If
+ * this doesn't sounds like what you want to parse have a look at
+ * {@link ObjectParser#declareNamedObjects(BiConsumer, ObjectParser.NamedObjectParser, Consumer, ParseField)} which solves a slightly
+ * different but similar sounding problem.
+ * <p>
+ * Anyway, {@linkplain ObjectSpooler} parses the fields in the order that they are in the XContent, collecting constructor arguments and
+ * "spooling" fields as they arrive by queueing the configured {@linkplain BiConsumer}, parsed value pair. Once all constructor arguments
+ * arrive it builds the target object and "unspools" the fields by draining the queue. Any fields that come in after the last constructor
+ * arguments are parsed just like {@linkplain ObjectParser}.
+ * </p>
+ * <p>
+ * Here is how you declare an ObjectSpooler:
+ * </p>
+ * <pre>{@code
+ *   private static final ObjectSpooler<Thing, Void> PARSER = new ObjectSpooler<>("thing",
+ *           a -> new Thing((String) a[0], (String) a[1]));
+ *   static {
+ *       PARSER.declareString(constructorArg(), new ParseField("animal"));
+ *       PARSER.declareString(constructorArg(), new ParseField("vegetable"));
+ *       PARSER.declareInt(Thing::setMineral, new ParseField("mineral"));
+ *       PARSER.declareInt(Thing::setFruit, new ParseField("fruit"));
+ *   }
+ * }</pre>
+ * <p>
+ * This does add some overhead compared to just using {@linkplain ObjectParser} directly. On a 2.2 GHz Intel Core i7 MacBook Air running on
+ * battery power in a reasonably unscientific microbenchmark it is about 100 microseconds for a reasonably large object, less if the
+ * constructor arguments are first. On this platform with the same microbenchmarks just creating the XContentParser is around 900
+ * microseconds and using {#linkplain ObjectParser} directly adds another 300 or so microseconds. In the best case
+ * {@linkplain ObjectSpooler} allocates two additional objects per parse compared to {#linkplain ObjectParser}. In the worst case it
+ * allocates <code>2 + 2 + (2 * param_count)</code> objects per parse, plus a few ArrayList resizes for very long lists. If this overhead
+ * is too much for you then feel free to have ObjectParser parse a secondary object and have that one call the target object's constructor.
+ * </p>
+ */
+public final class ObjectSpooler<Value, Context> extends AbstractObjectParser<Value, Context> {
```

Comment:
I don't like the name `ObjectSpooler`. My biggest issue with it is that it's completely missing the fact that this is a parser but also it doesn't give a big indication that this is for parsing objects that have required arguments. I don't have a good suggestion, but hopefully my reasons for not liking it will help come up with something better?